### PR TITLE
[#327] Generate diagnostics when a file is opened

### LINKED ISSUE #327: Generate diagnostics when a file is opened
**Is your feature request related to a problem? Please describe.**

At the moment, diagnostics are only generated when a file is saved.

**Describe the solution you'd like**

When a file is opened, the diagnostics should be provided.

### PR DESCRIPTION
**Pull request by alanz**
Fixes #327.

### CODE REVIEW

**Comment by alanz** [comment]
I am seeing something in my log that was not there before this commit

```
2019-12-17 16:13:13.510 [warning] <0.169.0> ** Undefined handle_info in els_server
** Unhandled message: {'EXIT',<0.657.0>,normal}
2019-12-17 16:13:41.781 [warning] <0.169.0> ** Undefined handle_info in els_server
** Unhandled message: {'EXIT',<0.693.0>,normal}
2019-12-17 16:15:31.815 [warning] <0.169.0> ** Undefined handle_info in els_server
** Unhandled message: {'EXIT',<0.695.0>,normal}
2019-12-17 16:15:33.291 [warning] <0.169.0> ** Undefined handle_info in els_server
** Unhandled message: {'EXIT',<0.696.0>,normal}
2019-12-17 16:22:01.893 [warning] <0.169.0>@els_methods:not_implemented_method:72 [Method not implemented] [method=$/cancelRequest]
2019-12-17 16:22:04.980 [warning] <0.169.0> ** Undefined handle_info in els_server
** Unhandled message: {'EXIT',<0.731.0>,normal}
2019-12-17 16:22:08.756 [warning] <0.169.0> ** Undefined handle_info in els_server
** Unhandled message: {'EXIT',<0.732.0>,normal}
```
Any idea where/how it could be coming about?

File: src/els_text_synchronization.erl [hunk 1]
```diff
@@ -1,5 +1,7 @@
 -module(els_text_synchronization).
 
+-include("erlang_ls.hrl").
```
Comment (jfacorro) [inline src/els_text_synchronization.erl]: Why was this added?

**Comment by jfacorro** [comment]
@alanz I'm not sure where those message are coming from, but the function `els_text_synchronization:did_save/1` is called with a spawn from the `els_methods:textdocument_didsave/2` function, I think we should do the same (at least for generating the diagnostics) when we call the `generate_diagnostics/1` function during `did_open/1`.

File: src/els_text_synchronization.erl [hunk 2]
```diff
@@ -11,12 +13,24 @@ did_open(Params) ->
...
+
+-spec did_close(map()) -> ok.
+did_close(_Params) -> ok.
+
+%%==============================================================================
+%% Internal functions
+%%==============================================================================
+
+-spec generate_diagnostics(uri()) -> ok.
+generate_diagnostics(Uri) ->
```
Comment (jfacorro) [inline src/els_text_synchronization.erl]: I'm not sure where those message are coming from. But the function `els_text_synchronization:did_save/1` is spawned from the `els_methods:textdocument_didsave/2` function. I think we should do the same (at least for generating the diagnostics) when we call the `generate_diagnostics/1` function during `els_text_synchronization:did_open/1`.

Comment (robertoaloi) [reply src/els_text_synchronization.erl]: I agree. This should be done asynchronously.

File: src/els_text_synchronization.erl (hunk 1 above)
Comment (alanz) [reply src/els_text_synchronization.erl]: Yes, I added it to fix compilation.

Aside: we should check that includes *are* actually needed.

File: src/els_text_synchronization.erl (hunk 2 above)
Comment (alanz) [reply src/els_text_synchronization.erl]: Great. I thought it would make sense to do that, but wanted feedback.

Comment (alanz) [reply src/els_text_synchronization.erl]: Will do.

File: src/els_text_synchronization.erl [hunk 3]
```diff
@@ -1,22 +1,39 @@
...
-  ok           = els_indexer:index(Document).
+  ok           = els_indexer:index(Document),
+  spawn (?MODULE, generate_diagnostics, [Uri]),
+  ok.
 
 -spec did_save(map()) -> ok.
 did_save(Params) ->
   TextDocument = maps:get(<<"textDocument">>, Params),
   Uri          = maps:get(<<"uri">>         , TextDocument),
+  generate_diagnostics(Uri).
```
Comment (jfacorro) [inline src/els_text_synchronization.erl]: Sorry I didn't mentioned it before but I think it would make more sense if we also used the spawn here instead of in `els_methods/didsave/2` function.

**Comment by alanz** [comment]
One thing that worries me slightly is that we are not ensuring that we only have one compile running per file at a time.  Given the stateful interaction with the file system, this is probably something we should serialise. 

This will  become more of a concern when #329 gets resolved and we run dialyzer on save, as it potentially takes a long time to run too.
